src/main/java/net/neoforged/neoforge/client/model/generators/template/ExtendedModelTemplateBuilder.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.model.generators.template;
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+import net.minecraft.client.data.models.model.ModelTemplate;
+import net.minecraft.client.data.models.model.TextureSlot;
+import net.minecraft.client.resources.model.UnbakedModel;
+import net.minecraft.resources.ResourceLocation;
+import net.neoforged.neoforge.client.event.RegisterNamedRenderTypesEvent;
+import org.jetbrains.annotations.Nullable;
+
+public class ExtendedModelTemplateBuilder {
+    Optional<ResourceLocation> parent = Optional.empty();
+    Optional<String> suffix = Optional.empty();
+    final Set<TextureSlot> requiredSlots = new HashSet<>();
+    final TransformsBuilder transforms = new TransformsBuilder();
+    final List<ElementBuilder> elements = new ArrayList<>();
+    @Nullable
+    CustomLoaderBuilder customLoader = null;
+    final RootTransformsBuilder rootTransforms = new RootTransformsBuilder();
+    @Nullable
+    ResourceLocation renderType = null;
+    @Nullable
+    Boolean ambientOcclusion = null; // UnbakedModel.DEFAULT_AMBIENT_OCCLUSION
+    @Nullable
+    UnbakedModel.GuiLight guiLight = null;
+
+    public static ExtendedModelTemplateBuilder of(ModelTemplate template) {
+        ExtendedModelTemplateBuilder builder = new ExtendedModelTemplateBuilder();
+        builder.parent = template.model;
+        builder.suffix = template.suffix;
+        builder.requiredSlots.addAll(template.requiredSlots);
+        if (template instanceof ExtendedModelTemplate ext) {
+            builder.transforms.copyFrom(ext.transforms);
+            ext.elements.forEach(elem -> builder.elements.add(elem.copy()));
+            builder.customLoader = ext.customLoader != null ? ext.customLoader.copy() : null;
+            builder.rootTransforms.copyFrom(ext.rootTransforms);
+            builder.renderType = ext.renderType;
+            builder.ambientOcclusion = ext.ambientOcclusion;
+            builder.guiLight = ext.guiLight;
+        }
+        return builder;
+    }
+
+    public static ExtendedModelTemplateBuilder builder() {
+        return new ExtendedModelTemplateBuilder();
+    }
+
+    public ExtendedModelTemplateBuilder parent(ResourceLocation parent) {
+        this.parent = Optional.of(parent);
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder suffix(String suffix) {
+        this.suffix = Optional.of(suffix);
+        return this;
+    }
+
+    public ExtendedModelTemplateBuilder requiredTextureSlot(TextureSlot slot) {
+        this.requiredSlots.add(slot);
+        return this;
+    }
```

Comment:
These methods need some short doc comments about what they are for, particularly the suffix and texture slot ones as they have a significant impact on the serialization behaviour of the template.